core/unix/os.c

```diff
@@ -6375,6 +6399,14 @@ handle_exit(dcontext_t *dcontext)
     }
     KSTOP(num_exits_dir_syscall);
 
+    /* This thread is on its way to exit, we are blocking all signals since any
+     * signal that reaches us now can be delayed until after the exit is complete.
+     * We may still receive a suspend signal for synchronization that we may need
+     * to reply to. If we are doing a detach, the thread's sigmask will be swapped
+     * back to the app's mask in signal_thread_exit (i#2921).
+     */
+    block_all_signals_except(NULL, 1, SUSPEND_SIGNAL);
```

Comment:
Done

Reply:
Could this go into one of the *_thread_exit() routines?

Reply:
It could, only problem, we need to know whether it was terminate_via_kill, i.e. if terminating syscall is SYS_kill, so we had to pass this into the thread_exit function(s). This is why I didn't do it there.